static-analysis/teammates-macker.xml

```diff
@@ -26,20 +181,57 @@
             </deny>
         </access-rule>
     </ruleset>
-    <ruleset name="Only UI tests can access page object classes">
+    <ruleset name="TestDriverComponent: Only UI tests can access page object classes">
         <access-rule>
             <deny>
                 <from class="teammates.test.**" />
-                <to class="teammates.test.pageobjects.*" />
+                <to class="teammates.test.pageobjects.**" />
                 <allow>
                     <!-- allow dependency within the package -->
-                    <from class="teammates.test.pageobjects.*" />
+                    <from class="teammates.test.pageobjects.**" />
                 </allow>
                 <allow>
                     <!-- allow dependency from UI tests -->
-                    <from class="teammates.test.cases.browsertests.*" />
+                    <from class="teammates.test.cases.browsertests.**" />
                 </allow>
             </deny>
         </access-rule>
     </ruleset>
+    <ruleset name="TestDriverComponent: Test driver can only be accessed from test package">
+        <access-rule>
+            <deny>
+                <from class="teammates.**" />
+                <to class="teammates.test.driver.**" />
+                <allow>
+                    <from class="teammates.test.**" />
+                </allow>
+                <allow>
+                    <from class="teammates.client.**" />
+                </allow>
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="TestDriverComponent: util and browsertest cannot access GaeSimulation">
```

Comment:
You should actually enforce the other way around: only certain test cases can access `GaeSimulation`. Namely, they are in the packages `cases.storage`, `cases.logic`, `cases.action`, `cases.automated`, `cases.search`.

Leaving a sneak preview that you're going to encounter one violation.